### PR TITLE
Elementwise multiplication with broadcasting

### LINKED ISSUE #557: Elementwise multiplication with broadcasting

### PR DESCRIPTION
**Pull request by soumyac1999**
**Description of changes:**

**Related Issues:**

Linked Issues:
- Issue #557 

Issues closed by this PR:
- Closes #557

**Before merging:**

- [ ] Did you update the [flexflow-third-party](https://github.com/flexflow/flexflow-third-party) repo, if modifying any of the Cmake files, the build configs, or the submodules?

### CODE REVIEW

**Comment by soumyac1999** [comment]
@jiazhihao I am trying to implement broadcasting for elementwise multiplication. I think I've got the forward pass. I had some questions about the backward pass. I think the computation will be as:
```
in1_grad = in1_grad + reduce(out_grad * in2)
and similarly for in2_grad
```

Which I think would require storing `out_grad * in2` in a temporary variable. And maybe another one for `in2_grad` due to the different shapes. But, these two temporary variables would not be used for the other elementwise binary operations and probably cause some memory overhead.

Would there be some simpler method to implement it without requiring so many changes? Or will we have to add in two temporary variables?

cc: @lockshaw

File: src/ops/kernels/element_binary_kernels.cu [hunk 1]
```diff
@@ -266,29 +266,68 @@ void forward_kernel(ElementBinaryMeta const *m,
...
-                             &alpha2,
-                             m->input2Tensor,
-                             in2_ptr,
-                             &alpha1,
-                             m->outputTensor,
-                             out_ptr));
+    assert(m->op_type == OP_EW_SUB || m->op_type == OP_EW_ADD ||
+           m->op_type == OP_EW_MUL);
+    if(m->op_type == OP_EW_SUB || m->op_type == OP_EW_ADD){
+      // output = (beta*output + alpha1*input1) + beta*output = input1
```
Comment (soumyac1999) [inline src/ops/kernels/element_binary_kernels.cu:272]: Are these comments useful? If not, will remove then

Comment (jiazhihao) [reply src/ops/kernels/element_binary_kernels.cu:272]: > @jiazhihao I am trying to implement broadcasting for elementwise multiplication. I think I've got the forward pass. I had some questions about the backward pass. I think the computation will be as:
> 
> ```
> in1_grad = in1_grad + reduce(out_grad * in2)
> and similarly for in2_grad
> ```
> 
> Which I think would require storing `out_grad * in2` in a temporary variable. And maybe another one for `in2_grad` due to the different shapes. But, these two temporary variables would not be used for the other elementwise binary operations and probably cause some memory overhead.
> 
> Would there be some simpler method to implement it without requiring so many changes? Or will we have to add in two temporary variables?

The backward equation you showed is correct, and I think you are right that we need a temporary buffer for saving out_grad * in2. An easy way to create it is use part of workspace. Note that workspace is allocated for operators that require saving intermediate buffers (e.g., convolutions, and this one). You can reserve part of the workspace for saving  `out_grad * in2` and the remaining for `cudnnReduceTensor`, which also requires a workspace (https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnReduceTensor)

**Comment by soumyac1999** [comment]
@jiazhihao, thanks for the reply! For now I got it to work my using a temporary variable as
```cpp
        float *temp_ptr = (float *)malloc(sizeof(*out_grad_ptr));
        checkCUDA(cudaMalloc(&temp_ptr, sizeof(*out_grad_ptr)));
```

Is this okay or is using the workspace the preferred method? If the latter, I will try using the workspace.

**Comment by jiazhihao** [comment]
We may want to use workspace since the cudaMalloc you add may cause out-of-memory issues. If you are not sure how to use the workspace for this, you can use cudaMalloc in this PR and create a github issue after merging the PR. I will push another fix for using workspace instead of cudaMalloc.

**Comment by soumyac1999** [comment]
@jiazhihao, I stored `out_grad * in2` in `m->handle.workSpace` and used `m->handle.workSpace + sizeof(*out_grad_ptr)` as the workspace for `cudnnReduceTensor` (with size `m->handle.workSpaceSize - sizeof(*out_grad_ptr)`). Is this the correct way to do it?

**Review by jiazhihao — APPROVED** [review]

**Pull request merged**